docs/routing.md

```diff
@@ -0,0 +1,266 @@
+- Return the final response.
+
+In order to prevent runaway phase checking, i.e. infinitely looping through phases, we have checks that determine whether we have processed phases too many times. This is indicative of records in the build output configuration causing infinite loops, and the counter resets each time the router is run for a request.
+
+### Checking Phases
+
+When we receive a request, the router looks for matches in the different phases, starting at the `none` phase.
+
+For every single source route in a phase, we check to see whether the route is a match, and process the relevant details (for more details on source checking see the [Checking Source Routes section](#checking-source-routes)). When we receive a response from the route checker, if we encountered an error — normally from running middleware — we break out of the routing process and return the error response. Otherwise, the only other time we disable the continuation of routing is if the route checker returns a final match.
+
+After all the source routes in the current phase have been checked, we move on to determine what will happen next. If the current phase is the `hit` phase, or the route has at some point been rewritten to an URL, the routing is complete and we're ready to server the the response.
+
+If the current phase is `miss`, we check the build output, if there is no file matching the requested one we set the status code to 404. Afterwards, if a file was found in the output, or the current phase is `error`, we move on to the `hit` phase. It is important that the hit phase is run every time a match is found in the build output or when the current phase is `miss` or `error`, because this allows us to update the headers for the response with the current matched path.
+
+In all other cases, we attempt to retrieve the next phase, based on the current one, and iterate forward.
+
+#### Diagram
+
+```mermaid
+flowchart TD
+    checkPhase --for each route--> checkRoute
+    checkRoute --> CheckRouteResult{check result}
+    CheckRouteResult -->|error| checkPhase
+    CheckRouteResult -->|done| IsPhaseHit{phase == hit}
+    CheckRouteResult -->|skip or next| CheckRouteResult
+    IsPhaseHit --> |yes| EndIsPhaseHit[matching done]
+    IsPhaseHit --> |no| IsPhaseMiss{phase == miss}
+    IsPhaseMiss --> |yes| Set404[set 404]
+    Set404 --> NextPhaseIsWhat{path found\nor\nphase == miss/error}
+    IsPhaseMiss --> |no| NextPhaseIsWhat
+    NextPhaseIsWhat --> |if true\nset next phase to hit| checkPhase
+    NextPhaseIsWhat --> |if false\nget next phase| getNextPhase --> checkPhase
+
+    NextPhaseIs{determine next phase} -->
+    getNextPhase --> NextPhaseIs{determine next phase}
+
+    NextPhaseIs --> |none| NextIsFS[filesystem]
+    NextPhaseIs --> |filesystem| NextIsRW[rewrite]
+    NextPhaseIs --> |rewrite| NextIsRS[resource]
+    NextPhaseIs --> |resource| NextIsMS[miss]
+    NextPhaseIs --> |all others| NextIsMS
+```
```

Comment:
I've been trying to improve this as the diagram fells too much complex/big to me, but it's a bit tricky 😓

Could something like this work?
![Screenshot 2023-05-11 at 23 46 19](https://github.com/cloudflare/next-on-pages/assets/61631103/5ca4a725-08d4-43b7-9760-697fbdae2111)

I _think_ that the rest of the info is present in the md already so here we can simplify things a bit?

```suggestion
```mermaid
flowchart TD
    checkPhase[check phase] --for each route--> checkRoute
    checkRoute[check route] -->|error| checkPhase
    checkRoute -->|done| phaseHitMiss{is phase\nhit or miss?}
    phaseHitMiss --> |phase == hit| matchingDone[matching done]
    phaseHitMiss --> |phase == miss| set404[set 404]
    phaseHitMiss --> |phase not neither| nextPhaseIsWhat{path found\nor\nphase == miss/error}
    set404 --> nextPhaseIsWhat{path found\nsuccessfully?}
    nextPhaseIsWhat --> |true\nset next phase to hit| checkPhase
    nextPhaseIsWhat --> |false\nset next phase| checkPhase
```
```